Documentation for Automate-deployed on-prem Builder. (#2315)
chef/automate: components/automate-chef-io/content/docs/on-prem-builder.md

{{< /warning >}}

### System requirements
Currently, Habitat Builder requires Chef Automate to be installed on the same host for
authentication purposes. Outside the boundaries of a proof-of-concept exercise, we
recommend against running installations of Chef Automate and Habitat Builder on
the same host.

#### Hardware
The minimum with which Chef Automate and Habitat Builder can be deployed on a single host
is:

* 16 GB of RAM.
* 130 GB of disk space, available to /hab.
* 4 vCPUs.
* Inbound network connectivity from LAN (HTTP/HTTPS) is required for internal clients to access the on-premises Chef Habitat Builder.

For deployments that are expected to see production-scale workload, we recommend:

* 64 GB of RAM.
* 200 GB of diskspace, available to /hab.
* 16 vCPUs.

Roughly 80 GB of the disk space is designated for Chef Automate; the rest is used for
Habitat Builder and the artifacts it stores. The current implementation uses Minio for
Habitat artifact storage; we do not support using Artifactory for artifact storage.

#### Operating system
Chef Automate and Habitat Builder require

* a Linux kernel of version 3.2 or greater
* `systemd` as the init system
* `useradd`

> * `curl` or `wget`

Reviewer: [minor] I wonder if we can get rid of the curl/wget requirement now that we aren't using install.sh

Reviewer 2: How are we getting the package if not `curl https://packages.chef.io/files/current/latest/chef-automate-cli/chef-automate_linux_amd64.zip | gunzip - > chef-automate && chmod +x chef-automate`?

Make sure the directory `<artifact-dir>` and its contents are on a host that can
access your on-premises Habitat Builder at `https://<your_automate_fqdn>/bldr`.

1. If your on-premises Habitat Builder is using a self-signed certificate, copy the SSL
   public key certificate chain into `/hab/cache/ssl` on the host from which you will
   perform the bulk upload to Habitat Builder.
1. Run the `bulkupload` command to upload artifacts from `<artifact-dir>` to the
   `<channel>` channel in on-premises Habitat Builder:

   ```shell
   HAB_AUTH_TOKEN=<your_on-prem_Builder_instance_token> hab pkg bulkupload --url https://<your_automate_fqdn>/bldr/v1 --channel <channel> <artifact-dir> --auto-create-origins
   ```
   The `--auto-create-origins` flag will create each origin listed in the
   `<artifact-dir>/artifacts` directory. If you omit the `--auto-create-origins` flag,
   use the Habitat Builder UI to create the necessary origins before running the
   `bulkupload` command.

### Backup Habitat Builder
Builder backups are performed as part of the [Chef Automate backup process]({{<relref "backup.md">}}).

### Not currently supported
* high-availability/DR/multinode Builder

> 

Reviewer 3: Would be great to have a Troubleshooting section, that shows the admin how to turn on Debug logging for the builder services (similar to the current on-prem docs)

Author: Added a section on Operating Habitat Builder that points to the relevant documentation for Chef Automate.

```shell
HAB_AUTH_TOKEN=<your_public_Builder_instance_token> hab pkg download --target x86_64-linux --channel stable --file package_seed_lists/effortless_x86_64-linux_stable --download-directory builder_bootstrap
```
#### Bulk-upload packages to on-premises Habitat Builder
If your on-premises Habitat Builder installation is airgapped, move the directory `<download-dir>`
and its contents from the previous step to a host that can reach your on-premises Habitat Builder.

1. If your on-premises Habitat Builder's SSL certificate was issued from an internal
   Public Key Infrastructure, copy the SSL public key certificate chain into
   `/hab/cache/ssl` on the host from which you will upload packages to your on-premises
   Habitat Builder.
1. Inspect the contents of your `<download-dir>/artifacts` directory. For each origin
   listed in that directory, make sure that origin exists in your on-premises Habitat
   Builder. If that origin does not exist in your on-premises Habitat Builder, create it
   in the Habitat Builder UI.
1. Run the `bulkupload` command to upload artifacts from `<download-dir>` to the
   `<channel>` channel in on-premises Habitat Builder:

   ```shell

>    HAB_AUTH_TOKEN=<your_on-prem_Builder_instance_token> hab pkg bulkupload --url https://automate.example.com/bldr --channel <channel> <download-dir>

Author: URL was a guess; need to verify.

Download and unzip the installer:

```shell
curl https://packages.chef.io/files/current/latest/chef-automate-cli/chef-automate_linux_amd64.zip | gunzip - > chef-automate && chmod +x chef-automate
```

### Deploy Chef Automate and Chef Habitat Builder On-prem

Deploying Chef Habitat Builder with Chef Automate requires a Chef Automate license. If you
already have a Chef Automate license, you may use it for the deployment. Otherwise, you
can accept the 30-day trial license when you first sign in to Chef Automate.

If you are deploying Chef Habitat Builder with Chef Automate in an airgapped environment,
follow [the documentation on building an airgap bundle]({{< relref "airgapped-installation.md" >}}).

To deploy Chef Automate and Chef Habitat Builder, specify both the `builder` and `automate`
products on the command line. For example:

```shell
 ./chef-automate deploy --product builder --product automate
```

Accept the license with `y`.

### Sign in to Chef Automate and the Chef Habitat Builder

1. View your login credentials in the terminal with:

   ```bash
   cat automate-credentials.toml
   ```

    You should see output similar to:

   ```output

>    url = "https://<your_automate_fqdn>"

Reviewer 2: Since we aren't giving customization information in this install doc, we might want consider using the default `chef-automate.test` instead of `your_automate_fqdn`?

Author: The most popular default across in the docs is {{< example_fqdn "automate" >}}, so I used that.

```shell
git clone https://github.com/habitat-sh/on-prem-builder.git
```

Once the repository is successfully cloned, move into the `on-prem-builder` repository:

```shell
cd on-prem-builder
```

The Chef Automate installer uses a self-signed certificate. Copy the SSL public key certificate chain from Chef Automate into `/hab/cache/ssl` with this command:

```shell
cp /hab/svc/automate-load-balancer/data/<your_automate_fqdn>.cert /hab/cache/ssl/ssl-certificate.crt
```

### Download Seed List from the Public Chef Habitat Builder

Your host must have access to the internet to download the curated seed list packages from the **public** [Chef Habitat Builder](https://bldr.habitat.sh).
If you have not already done so, create a user account and personal access token on the **public** [Chef Habitat Builder](https://bldr.habitat.sh/).

Use the `hab pkg download` command with a seed list `</path/to/seed_list>` to download packages for
your desired architecture `<arch>` from a channel `<channel>` to a directory `<artifact-dir>`:

```shell

> HAB_AUTH_TOKEN=<our_public_builder_personal_access_token> hab pkg download --target <arch> --channel <channel> --file </path/to/seed_list> --download-directory <artifact-dir>

Reviewer 2: ```suggestion
HAB_AUTH_TOKEN=<your_public_builder_personal_access_token> hab pkg download --target <arch> --channel <channel> --file </path/to/seed_list> --download-directory <artifact-dir>
```

Found 46 artifact(s) for upload.
Discovering origin names from local artifact cache
Missing origin 'chef'
Origin 'core' already exists
Missing origin 'effortless'
Creating origin chef.
Created origin chef.
Creating origin effortless.
Created origin effortless.
75 B / 75 B | [===========================================] 100.00 % 1.31 MB/s d
Uploading public origin key chef-20160614114050.pub
...
```

The `--auto-create-origins` flag creates each origin listed in the
`<artifact-dir>/artifacts` directory. If you omit the `--auto-create-origins` flag,
use the Chef Habitat Builder UI to create the necessary origins before running the
`bulkupload` command.

To finish up, return to your Chef Habitat Builder on-prem installation and view the packages that you've added to your `core` origin at `https://<your_automate_fqdn>/bldr/#/origins/core/packages`.

## Operating Chef Habitat Builder

Chef Habitat Builder uses the same mechanisms that Chef Automate does for [backups]({{< relref "backup.md" >}}), [log management]({{< relref "log-management.md" >}}), and [uninstalling]({{< relref "troubleshooting.md#uninstalling-chef-automate" >}}).

> To change the log level for Chef Habitat Builder only, create a TOML file that contains the partial configuration below. Uncomment and change settings as needed, and then run `chef-automate config patch </path/to/your-file.toml>` to deploy your change.

Reviewer 2: ```suggestion
### Logging Errors

To change the log level for Chef Habitat Builder only, create a TOML file that contains the partial configuration below. Uncomment and change settings as needed, and then run `chef-automate config patch </path/to/your-file.toml>` to deploy your change.
```

* Effortless: packages used to start with the [Effortless pattern](https://github.com/chef/effortless). A complete Effortless implementation requires the contents of both the `stable` and the `unstable` channel.

### Clone the Chef Habitat Builder On-prem Repository

To access the curated seed lists for bootsrapping Chef Habitat Builder on-prem, you will need to clone the Chef Habitat Builder on-prem repository using https.

```shell
git clone https://github.com/habitat-sh/on-prem-builder.git
```

Once the repository is successfully cloned, move into the `on-prem-builder` repository:

```shell
cd on-prem-builder
```

The Chef Automate installer uses a self-signed certificate. Copy the SSL public key certificate chain from Chef Automate into `/hab/cache/ssl` with this command:

```shell
cp /hab/svc/automate-load-balancer/data/chef-automate.test.cert /hab/cache/ssl/ssl-certificate.crt
```

### Download Seed List from the Public Chef Habitat Builder

Your host must have access to the internet to download the curated seed list packages from the **public** [Chef Habitat Builder](https://bldr.habitat.sh).

> If you have not already done so, create a user account and personal access token on the **public** [Chef Habitat Builder](https://bldr.habitat.sh/)

Reviewer 4: ```suggestion
If you have not already done so, create a user account and personal access token on the **public** [Chef Habitat Builder](https://bldr.habitat.sh/).
```

{{< warning >}}
These instructions have been deprecated in favor of using the Chef Automate installer to deploy Chef Habitat on-prem
{{< /warning >}}

To configure Chef Automate as an OAuth Provider for Habitat Builder, create a TOML file with the partial configuration below.
Run `chef-automate config patch </path/to/your-file.toml>` to deploy your change.

`bldr_client_id` and `bldr_client_secret` simply need to match what you configured for the corresponding
values in Habitat Builder (see below). However, we strongly recommend those values follow
[best practices](https://www.oauth.com/oauth2-servers/client-registration/client-id-secret/)
for `client_id` and `client_secret` in the Oauth2 standard.

```toml
[session.v1.sys.service]
bldr_signin_url = "<your Builder fqdn>" # for example, "http://builder.test/"

# This needs to match what you configured OAUTH_CLIENT_ID as when you configured Habitat Builder.
bldr_client_id = "<your Habitat Builder Oauth2 Client ID>"

# This needs to match what you configured OAUTH_CLIENT_SECRET as when you configured Habitat Builder.
bldr_client_secret = "<your Habitat Builder Oauth2 Client Secret>"
```

> You'll need to add Automate's TLS certificate to Builder's list of accepted certificates

Reviewer 4: ```suggestion
In addition, add Automate's TLS certificate to Builder's list of accepted certificates.
```

> in addition to these configuration changes.

Reviewer 4: ```suggestion
```

> Locate Automate's default self-signed certificate by running `cat /hab/svc/automate-load-balancer/data/{{< example_fqdn "automate" >}}.cert`

Reviewer 4: ```suggestion
Locate Automate's default self-signed certificate by running `cat /hab/svc/automate-load-balancer/data/{{< example_fqdn "automate" >}}.cert`, copy this default certificate, and then add it to your Builder instance's list of accepted certificates.
```

> You can copy this default certificate, and then add it to your Builder instance's list of accepted certs.

Reviewer 4: ```suggestion
```

Reviewer 5: @IanMadd why are you suggesting that we remove this line?

+++
title = "Deploying Builder with Chef Automate"
description = "use Chef Automate to deploy Builder"
date = 2019-11-19T14:10:15-08:00
weight = 20
draft = false
bref = ""
toc = true
+++

## Overview
Enterprise customers may wish to set up an on-premises Habitat Builder depot to store
Habitat packages for use by their own Habitat Studios and Supervisors. Use Chef Automate
to deploy and manage an on-premises Habitat Builder depot.

{{< warning >}}
The current implementation of Automate-deployed Habitat Builder is an early milestone
intended for use in proof-of-concept deployments. We do not consider it a fully-mature

> feature, and you should not either.

Reviewer 6: :100:

{{< /warning >}}

### System requirements

> <!-- Question: What do we tell them about disk space requirements for Builder artifact

Reviewer 6: https://github.com/habitat-sh/on-prem-builder/blob/master/on-prem-docs/content/docs/getting-started.md#requirements

Reviewer: We should sync up those recommendations with the preflight checks

storage? -->

Currently, Habitat Builder requires Chef Automate to be installed on the same host for
authentication purposes. Outside the boundaries of a proof-of-concept exercise, we
recommend against running production installations of Chef Automate and Habitat Builder on
the same host.

### Deploy on-premises Habitat Builder
To deploy Chef Automate and Habitat Builder, run the command

```shell
chef-automate deploy --product builder --product automate
```

<!--
Questions:

> 1. If someone has an existing Chef Automate installation with the Builder bits, would it

Reviewer 6: I'm not sure what you're asking here. Right now you _can't_ run a builder and Automate because we'll snipe any hab services that aren't automate in the converge loop. Perhaps you're asking about somebody enabling builder on an existing Automate, and yeah, a config patch with the correct products _should_ do what need it to do.

Author: The question I was trying to ask was: if I already have Chef Automate installed, can/should I run `chef-automate deploy --product builder` on that installation to add builder?

The larger question is, what do we want to tell people who already have a recent-ish Chef Automate about how they can try Builder out?

   be safe for them to deploy the builder product in that installation?

> 2. How is licensing going to work?

Reviewer 6: Right now we require automate and automate requires a license. The result of https://github.com/chef/automate/issues/2248 and it's followups might be able to change that if we get rid of the automate product requirement though. I guess time will tell.

-->

> ### Login to Habitat Builder

Reviewer 7: both builder and automate now use _Sign in_, could we use that here as well for the heading and first step?

1. Login to Chef Automate using the <relref "install.md#open-chef-automate">credentials provided during deployment.</relfef>
1. Go to https://automate.example.com/bldr.

> 1. Click the Sign In button in the upper right corner.

Reviewer 7: we've been working to avoid using `Click` in the docs, as it implies everyone uses a pointing device. the docs standard is to use `Select` instad.

1. Click Sign In With Chef Automate.
1. Use the same credentials as in Step 1 to sign in to Habitat Builder.

See the <ref "https://www.habitat.sh/docs/using-builder/">Habitat Builder documentation</ref> for setting up a Builder access token, creating a Builder origin, etc.

The Automate-deployed Habitat Builder currently supports authentication with local users only.

### Backup Habitat Builder
<!-- Questions:

>   1. What is our commitment to forward compatibility of the backup format?

Reviewer 6: It's a proof-of-concept that we'll support forever ;)

Author: @ryancragun /me sighs deeply

Author: I'll add words about forward compatibility of the backup format.

>   2. Do we intend to support Builder-only backups? Right now, if someone wants to backup

Reviewer 6: Yeah, in practice those should be super small, but dealing with ElasticSearch for builder only deployment is :hankey: I'm firmly on the let's get it running without a hard-requirement-of-the-automate-product train.

will have the opportunity to download a trial license when you sign in to Chef Automate
after going through the deployment process.

If you are deploying Habitat Builder with Chef Automate in an airgapped environment,
follow <relref "airgapped-installation.md">the documentation on building an airgap bundle</relref>.

To deploy Chef Automate and Habitat Builder, specify both the `builder` and `automate`
products on the command line. For example,

```shell
sudo chef-automate deploy --product builder --product automate
```

### Sign in to Habitat Builder
1. Sign in to Chef Automate using the <relref "install.md#open-chef-automate">credentials provided during deployment.</relfef>
1. Go to https://automate.example.com/bldr.
1. Select the Sign In button in the upper right corner.
1. Select Sign In With Chef Automate.
1. Use the same credentials as in Step 1 to sign in to Habitat Builder.

See the <ref "https://www.habitat.sh/docs/using-builder/">Habitat Builder documentation</ref> for setting up a Builder access token, creating a Builder origin, etc.

The Automate-deployed Habitat Builder currently supports authentication with local users only.

### Backup Habitat Builder

> Builder backups are performed as part of the <relref "backup.md">Chef Automate backup process</relref>. The current implementation supports filesystem backups only.

Reviewer 8: i'm not sure this is true. I think S3 backups should just work

Author: @jaym Does #2393 look reasonable? If so, I will take out the part about filesystem backups only.

+++
title = "Deploying Builder with Chef Automate"
description = "use Chef Automate to deploy Builder"
date = 2019-11-19T14:10:15-08:00
weight = 20
draft = false
bref = ""
toc = true
+++

## Overview
Enterprise customers may wish to set up an on-premises Habitat Builder depot to store
Habitat packages for use by their own Habitat Studios and Supervisors. Use Chef Automate
to deploy and manage an on-premises Habitat Builder depot.

{{< warning >}}
The current implementation of Automate-deployed Habitat Builder is an early milestone
intended for use in proof-of-concept deployments. We do not consider it a fully-mature
feature, and you should not either.
{{< /warning >}}

### System requirements
Currently, Habitat Builder requires Chef Automate to be installed on the same host for
authentication purposes. Outside the boundaries of a proof-of-concept exercise, we
recommend against running installations of Chef Automate and Habitat Builder on
the same host.

#### Hardware
The minimum with which Chef Automate and Habitat Builder can be deployed on a single host
is:

> 

Author: The following numbers are from merging the numbers from

https://automate.chef.io/docs/system-requirements/

and 

https://github.com/habitat-sh/on-prem-builder/blob/master/on-prem-docs/content/docs/getting-started.md#requirements

If someone wants to suggest something else, that would be fine.

Author: I added #2353 for adjusting the Automate preflight checks to take Builder into account.

* 16 GB of RAM.
* 130 GB of disk space, available to /hab.
* 4 vCPUs.
* Inbound network connectivity from LAN (HTTP/HTTPS) is required for internal clients to access the on-premises Chef Habitat Builder.

For deployments that are expected to see production-scale workload, we recommend:
* 64 GB of RAM.
* 200 GB of diskspace, available to /hab.
* 16 vCPUs.

Roughly 80 GB of the disk space is designated for Chef Automate; the rest is used for
Habitat Builder and the artifacts it stores. The current implementation uses Minio for

> Habitat artifact storage; we do not yet support using Artifactory for artifact storage.

Author: ~yet~

#### Operating system
Chef Automate and Habitat Builder require

* a Linux kernel of version 3.2 or greater
* systemd as the init system
* useradd
* curl or wget
* The shell that starts Automate should have a max open files setting of at least 65535

> ### Deploy on-premises Habitat Builder

Author: We expect that most potential customers will be running this behind an airgap, so we should point to the directions for the airgap bundle.

Deploying Habitat Builder with Chef Automate requires a Chef Automate license. If you
already have a Chef Automate license, you may use it for the deployment. Otherwise, you
will have the opportunity to download a trial license when you sign in to Chef Automate
after going through the deployment process.

To deploy Chef Automate and Habitat Builder, run the command

```shell
chef-automate deploy --product builder --product automate
```

<!--
Question: Suppose someone has a recent-ish Chef Automate deployed and wants to try
Builder. What do we tell them to do?

> -->

Author: What do we want to tell a customer to do who has Chef Automate deployed and wants to try Builder? The options that come to my mind off the cuff are:

* config
* a command, e.g. `chef-automate deploy --product automate --product builder`
* tell them to deploy Automate + Habitat on a separate host

Author: Decision is to focus on the new installation and quietly not talk about this case.

+++
title = "Install Chef Habitat Builder On-prem"
description = "Deploy Chef Habitat Builder on-prem with the Chef Automate installer"
date = 2019-11-19T14:10:15-08:00
weight = 20
draft = false
bref = ""
toc = true
[menu]
  [menu.docs]
    parent = "get_started"
    weight = 50
+++

This guide details how to install Chef Automate and deploy Chef Habitat Builder on-prem together. Enterprise customers may wish to set up an on-premises Chef Habitat Builder depot to store Chef Habitat packages for use by their own Chef Habitat Studios and Supervisors.

{{< warning >}}
Automate-deployed Chef Habitat Builder is a new feature that is in active development. Please contact your Chef representative before using this implementation in production.
{{< /warning >}}

> This guide covers setting up Chef Automate and Chef Habitat Builder on-prem and bootstrapping Chef Habitat Builder on-prem with curated core seed lists from the Chef Habitat public Builder.

Reviewer 9: ```suggestion
This guide covers setting up Chef Automate and Chef Habitat Builder on-prem, and bootstrapping Chef Habitat Builder on-prem with curated core seed lists from the Chef Habitat public Builder.
```

* 16 vCPUs.

Roughly 80 GB of the disk space is designated for Chef Automate; the rest is used for
Chef Habitat Builder and the artifacts it stores. The current implementation uses Minio for
Chef Habitat artifact storage; we do not support using Artifactory for artifact storage.

### Operating System

Chef Automate and Chef Habitat Builder require:

* a Linux kernel of version 3.2 or greater
* `systemd` as the init system
* `useradd`
* `curl` or `wget`
* The shell that starts Automate should have a max open files setting of at least 65535

> * Run the installation and bootstrapping procedures as the superuser or use `` at the start of each command.

Reviewer 9: @kagarmoe Is there a command missing for ` `` `?  Asking as that throws off the formatting for this document.

→ Found 46 artifact(s) for upload.
☁ Discovering origin names from local artifact cache
∵ Missing origin 'chef'
✓ Origin 'core' already exists
∵ Missing origin 'effortless'
Ω Creating origin chef.
✓ Created origin chef.
Ω Creating origin effortless.
✓ Created origin effortless.
75 B / 75 B | [===========================================] 100.00 % 1.31 MB/s d
» Uploading public origin key chef-20160614114050.pub
...
```

The `--auto-create-origins` flag creates each origin listed in the
`<artifact-dir>/artifacts` directory. If you omit the `--auto-create-origins` flag,
use the Chef Habitat Builder UI to create the necessary origins before running the
`bulkupload` command.

To finish up, return to your Chef Habitat Builder on-prem installation and view the packages that you've added to your `core` origin at `https://chef-automate.test/bldr/#/origins/core/packages`.

## Operating Chef Habitat Builder

Chef Habitat Builder uses the same mechanisms that Chef Automate does for [backups]({{< relref "backup.md" >}}), [log management]({{< relref "log-management.md" >}}), and [uninstalling]({{< relref "troubleshooting.md#uninstalling-chef-automate" >}}).

> To change the log level for Chef Habitat Builder only, create a TOML file that contains the partial configuration below. Uncomment and change settings as needed, then run `chef-automate config patch </path/to/your-file.toml>` to deploy your change.

Reviewer 9: ```suggestion
To change the log level for Chef Habitat Builder only, create a TOML file that contains the partial configuration below. Uncomment and change settings as needed, and then run `chef-automate config patch </path/to/your-file.toml>` to deploy your change.
```

```toml
[builder_api.v1.sys.log]
level = "debug"
scoped_levels = ["tokio_core=error", "tokio_reactor=error", "zmq=error", "hyper=error" ]
```

## Setting up Automate as an OAuth Provider for Habitat Builder (Deprecated)

{{< warning >}}

> These instructions have been deprecated in favor of using the Chef Automate installer to deploy Chef Habitat on-prem

Reviewer 9: ```suggestion
These instructions have been deprecated in favor of using the Chef Automate installer to deploy Chef Habitat on-prem.
```

{{< /warning >}}

To configure Chef Automate as an OAuth Provider for Habitat Builder, create a TOML file with the partial configuration below.
Run `chef-automate config patch </path/to/your-file.toml>` to deploy your change.

`bldr_client_id` and `bldr_client_secret` simply need to match what you configured for the corresponding

> values in Habitat Builder (see below). However, we strongly recommend those values follow

Reviewer 9: ```suggestion
values in Chef Habitat Builder (see below). However, we strongly recommend those values follow
```

### Operating Habitat Builder
Habitat Builder uses the same mechanisms that Chef Automate does for [backups]({{<relref "backup.md">}}), [log management]({{<relref "log-management.md">}}), and [uninstallation]({{<relref "troubleshooting.md#uninstalling-chef-automate">}}).

To change the log level for Habitat Builder only, create a TOML file that contains the partial configuration below. Uncomment and change settings as needed, then run `chef-automate config patch </path/to/your-file.toml>` to deploy your change.

```toml
[builder_api.v1.sys.log]
level = "debug"
scoped_levels = ["tokio_core=error"]
```

> Set the log level to `error` for normal operation and to `debug` for any debugging. Due to

Reviewer 8: this is not correct. `level` is whatever automate accepts (`debug`, `info`, `warning`, `error`, `fatal`, `panic` I think). I would avoid mentioning scoped_levels entirely as its an implementation detail that we can set with support when needed

Author: @jaym I've removed `scoped_levels` from the config example and the incorrect text about levels.
